src/HomogeneousCalorimeter_geo.cpp

```diff
@@ -288,311 +184,153 @@ std::tuple<Volume, Position> build_module(Detector& desc, xml::Collection_t& plm
-                                      int sid)
-{
-  auto [modVol, modSize] = build_module(desc, plm, sens);
-  int sector_id          = dd4hep::getAttrOrDefault<int>(plm, _Unicode(sector), sid);
-  int id_begin           = dd4hep::getAttrOrDefault<int>(plm, _Unicode(id_begin), 1);
-  int nrow               = plm.attr<int>(_Unicode(nrow));
-  int ncol               = plm.attr<int>(_Unicode(ncol));
-
-  // compute array position
-  double begx = -modSize.x() * ncol / 2. + modSize.x() / 2.;
-  double begy = modSize.y() * nrow / 2. - modSize.y() / 2.;
-
-  std::vector<std::pair<int, int>> removals;
-  // get the removal list
-  for (xml::Collection_t rm(plm, _Unicode(removal)); rm; ++rm) {
-    removals.push_back({rm.attr<int>(_Unicode(row)), rm.attr<int>(_Unicode(col))});
-  }
 
-  // placement inside mother
-  auto pos = get_xml_xyz(plm, _Unicode(position));
-  auto rot = get_xml_xyz(plm, _Unicode(rotation));
+  //=========================================================
+  // The modules' positions followd the fillRectangles function
+  //=========================================================
+  float half_modx = modSize.x() * 0.5, half_mody = modSize.y() * 0.5;
+  auto points = epic::geo::fillRectangles({half_modx, half_mody}, modSize.x(), modSize.y(), rmin, rmax, phimin, phimax);
 
-  // optional envelope volume
-  bool        has_envelope = dd4hep::getAttrOrDefault<bool>(plm, _Unicode(envelope), false);
-  Material    material     = desc.material(getAttrOrDefault<std::string>(plm, _U(material), "Air"));
-  Box         solid(ncol * modSize.x() / 2.0, nrow * modSize.y() / 2.0, modSize.z() / 2.0);
-  Volume      env_vol(std::string(env.name()) + "_envelope", solid, material);
-  Transform3D tr_global = RotationZYX(rot.z(), rot.y(), rot.x()) * Translation3D(pos.x(), pos.y(), pos.z());
-  if (has_envelope) {
-    env.placeVolume(env_vol, tr_global);
-  }
 
-  // local placement of modules
-  int nmodules = 0;
-  for (int i = 0; i < nrow; ++i) {
-    for (int j = 0; j < ncol; ++j) {
-      if (std::find(removals.begin(), removals.end(), std::pair<int, int>(i, j)) != removals.end()) {
-        continue;
+  //=========================================================
+  // Read the positions information from xml file
+  //=========================================================
+  for (xml_coll_t x_positions_i(plm, _Unicode(positions)); x_positions_i; ++x_positions_i){
+    xml_comp_t x_positions = x_positions_i;
+
+    for (xml_coll_t x_position_i(x_positions, _U(position)); x_position_i; ++x_position_i){
+      xml_comp_t x_position = x_position_i;
+
+      if( x_positions.nameStr() == "extrudedpolygon_pos" ){  // retrieve the position of extrudedpolygon shape supporting structure
+        pt_innerframe_x.push_back((x_position.x())*cm);
+        pt_innerframe_y.push_back((x_position.y())*cm);
+      }
+      else{  // retrieve the position of additional modules
+        auto add_point = epic::geo::Point((x_position.x())*cm, (x_position.y())*cm);
```

Comment:
Related with adding dimensions like `cm` into the xml file, you would have to remove them from here. For physical quantities, always keep the values together with their dimension for clarity. This currently puts the dimension in the cpp file, but the values in the xml file.

Reply:
Got it! this is the same issue that I need to assign unit in the xml file. I remove this in my next commit. I will push it soon.